### PR TITLE
feat: add get_{ref, mut} to arrow_ipc Reader and Writer

### LINKED ISSUE #4121: Support accessing ipc Reader/Writer inner by reference
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Our own Reader passed into `arrow_ipc::reader::FileReader` includes statistics, and we need to collect statistics during iterating the reader. Therefore, we need to access the inner reader through `arrow_ipc::reader::FileReader`.

**Describe the solution you'd like**

Add `get_ref` and `get_mut` methods to ipc readers and writers.

**Describe alternatives you've considered**

Wrapping our own reader or writer with `Arc<Mutex>`.

**Additional context**

### PR DESCRIPTION
**Pull request by sticnarf**
# Which issue does this PR close?

Closes #4121.

# Rationale for this change
 

# What changes are included in this PR?

The `BufReader` and `BufWriter` in Rust `std` have similar methods.

# Are there any user-facing changes?

### CODE REVIEW

**Review by waynexia — APPROVED** [review]
Looking great 👍 thanks @sticnarf for your contribution

File: arrow-ipc/src/reader.rs [hunk 1]
```diff
@@ -1038,6 +1038,20 @@ impl<R: Read + Seek> FileReader<R> {
...
+    /// Gets a reference to the underlying reader.
+    ///
+    /// It is inadvisable to directly read from the underlying reader.
+    pub fn get_ref(&self) -> &R {
+        self.reader.get_ref()
+    }
+
+    /// Gets a mutable reference to the underlying reader.
+    ///
+    /// It is inadvisable to directly read from the underlying reader.
```
Comment (tustvold) [inline arrow-ipc/src/reader.rs:1051]: :+1:

File: arrow-ipc/src/reader.rs [hunk 2]
```diff
@@ -1038,6 +1038,20 @@ impl<R: Read + Seek> FileReader<R> {
             ))),
         }
     }
+
+    /// Gets a reference to the underlying reader.
+    ///
+    /// It is inadvisable to directly read from the underlying reader.
```
Comment (tustvold) [inline arrow-ipc/src/reader.rs:1044]:
```suggestion
```

Reading requires mutable access so this comment is possibly redundant

File: arrow-ipc/src/reader.rs [hunk 3]
```diff
@@ -1243,6 +1257,20 @@ impl<R: Read> StreamReader<R> {
             )),
         }
     }
+
+    /// Gets a reference to the underlying reader.
+    ///
+    /// It is inadvisable to directly read from the underlying reader.
```
Comment (tustvold) [inline arrow-ipc/src/reader.rs:1263]:
```suggestion
```

**Review by tustvold — APPROVED** [review]
This looks good to me, my only concern is that get_mut might cause users to rely on implicit behaviour of the outer types (e.g. when they flush, etc...) but the "inadvisable" comment hopefully protects us against this

File: arrow-ipc/src/reader.rs (hunk 2 above)
Comment (sticnarf) [reply arrow-ipc/src/reader.rs:1044]: Sometimes reading doesn't require a mutable reference (e.g. `&File` implements `Read`), so I think the comment is worth having.

**Pull request merged**